src/editor/Model/FileExplorer.re

```diff
@@ -0,0 +1,164 @@
+open Revery;
+open Oni_Core;
+
+type t = {
+  directory: UiTree.t,
+  isOpen: bool,
+};
+
+module ExplorerId =
+  UniqueId.Make({});
+
+let toFsNode = node =>
+  switch (node) {
+  | UiTree.FileSystemNode(n) => n
+  };
+
+let getFileIcon = (languageInfo, iconTheme, filePath) => {
+  let fileIcon =
+    LanguageInfo.getLanguageFromFilePath(languageInfo, filePath)
+    |> IconTheme.getIconForFile(iconTheme, filePath);
+
+  switch (fileIcon) {
+  | Some(_) as x => x
+  | None => None
+  };
+};
+
+let createFsNode = (~children, ~path, ~displayName, ~fileIcon, ~isDirectory) => {
+  /**
+     TODO: Find an icon theme with folders and use those icons
+     Fallbacks are used for the directory icons. FontAwesome is not
+     accessible here so we specify no icons for directories.
+   */
+  let (primary, secondary) = isDirectory ? (None, None) : (fileIcon, None);
+
+  UiTree.FileSystemNode({
+    path,
+    displayName,
+    children,
+    isDirectory,
+    icon: primary,
+    secondaryIcon: secondary,
+  });
+};
+
+let isDir = path => {
+  Sys.file_exists(path) ? Sys.is_directory(path) : false;
+};
+
+/**
+  getFilesAndFolders
+
+   This function uses Lwt to get all the files and folders in a directory
+   then for each we check if it is a file, if so we create a filesystem
+   node (an in-memory representation of the file) without children
+   if it is a directory we recursively call getFilesAndFolders on it
+   to resolves its subfolders and files. We do this concurrently using
+   Lwt_list.map_p
+
+   TODO: Add the concept of depth to this function so we can terminate
+   the file resolution early and avoid choking on large directories
+ */
+let rec getFilesAndFolders = (cwd, getIcon, ~ignored) => {
+  /* Wrap the operation in a try%lwt which will catch all async exceptions */
+  try%lwt (
+    Lwt_unix.files_of_directory(cwd)
```

Comment:
Neat! Good idea to use `Lwt` here 👍

Reply:
Will be nice for perf because the async I/O can be run in parallel with other threads (like the rendering).